server/src/main/java/eu/fasten/server/plugins/kafka/FastenKafkaPlugin.java

```diff
@@ -0,0 +1,363 @@
+        this.plugin = plugin;
+
+        this.connection = new KafkaConsumer<>(p);
+        this.serverLog = new KafkaProducer<>(
+                this.setKafkaProducer(p.getProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG),
+                        plugin.getClass().getSimpleName() + "_server_logs"));
+        this.producer = new KafkaProducer<>(
+                this.setKafkaProducer(p.getProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG),
+                        plugin.getClass().getSimpleName() + "_CGS_status"));
+
+        this.consumerHostName = this.getConsumerHostName();
+        this.skipOffsets = skipOffsets;
+
+        logger.debug("Constructed a Kafka plugin for " + plugin.getClass().getCanonicalName());
+    }
+
+    @Override
+    public void run() {
+        try {
+            if (plugin.consumeTopics().isPresent()) {
+                connection.subscribe(plugin.consumeTopics().get());
+            }
+            if (this.skipOffsets == 1) {
+                skipPartitionOffsets();
+            }
+
+            while (true) {
+                if (plugin.consumeTopics().isPresent()) {
+                    handleConsuming();
+                } else {
+                    doCommitSync();
+
+                    handleProducing(null, null);
+                }
+            }
+        } catch (WakeupException e) {
+            logger.debug("Caught wakeup exception");
+        } finally {
+            connection.close();
+        }
+    }
+
+    /**
+     * Starts a thread.
+     */
+    public void start() {
+        this.thread = new Thread(this);
+        this.thread.setName(this.plugin.getClass().getSimpleName() + "_plugin");
+        this.thread.start();
+    }
+
+    /**
+     * Sends a wake up signal to Kafka consumer and stops it.
+     */
+    public void stop() {
+        connection.wakeup();
+    }
+
+    /**
+     * Getter for the thread.
+     *
+     * @return thread
+     */
+    public Thread thread() {
+        return thread;
+    }
+
+    /**
+     * Consumes a message from a Kafka topics and passes it to a plugin.
+     */
+    private void handleConsuming() {
+        ConsumerRecords<String, String> records = connection.poll(Duration.ofSeconds(1));
+        for (var r : records) {
+            doCommitSync();
+
+            long startTime = System.currentTimeMillis();
+            plugin.consume(r.value());
+            long endTime = System.currentTimeMillis();
+
+            handleProducing(r.value(), String.valueOf(endTime - startTime));
+        }
+    }
+
+    /**
+     * Writes messages to server log and stdout/stderr topics.
+     *
+     * @param input input message [can be null]
+     * @param time  precessing time [can be null]
+     */
+    private void handleProducing(String input, String time) {
+        if (plugin.recordProcessSuccessful()) {
+            var result = plugin.produce();
+
+            emitMessage(this.serverLog, this.serverLogTopic,
```

Comment:
Why are we writing server log messages to Kafka and not to the standard log? If the purpose is debugging, then we need to make a choice: we either write _everything_ to Kafka (e.g. by using a Kafka-aware logger) or we use a standard log aggregator that collects all server instance logs to a centralised location.

Reply:
Moreover, this method cries for a `StringBuffer` to build the message string

Reply:
The format should be JSON again, other wise we are inconsistent.

Reply:
The only things that are written to Kafka are stdout and stderr now. Using logger for the rest